universal-application-tool-0.0.1/app/models/Version.java

```diff
@@ -0,0 +1,65 @@
+package models;
+
+import com.google.common.collect.ImmutableList;
+import io.ebean.annotation.UpdatedTimestamp;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import javax.persistence.Entity;
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import play.data.validation.Constraints;
+
+@Entity
+@Table(name = "versions")
+public class Version extends BaseModel {
+  @Constraints.Required private LifecycleStage lifecycleStage;
+
+  @ManyToMany(mappedBy = "versions")
+  private List<Question> questions;
+
+  @ManyToMany(mappedBy = "versions")
+  private List<Program> programs;
+
+  @UpdatedTimestamp private Instant submitTime;
+
+  public Version() {
+    this(LifecycleStage.DRAFT);
+  }
+
+  public Version(LifecycleStage lifecycleStage) {
+    this.lifecycleStage = lifecycleStage;
+  }
+
+  public void setLifecycleStage(LifecycleStage lifecycleStage) {
+    this.lifecycleStage = lifecycleStage;
+  }
+
+  public ImmutableList<Program> getPrograms() {
+    return ImmutableList.copyOf(programs);
+  }
+
+  public ImmutableList<Question> getQuestions() {
+    return ImmutableList.copyOf(questions);
+  }
+
+  public LifecycleStage getLifecycleStage() {
+    return lifecycleStage;
+  }
+
+  public Instant getSubmitTime() {
+    return this.submitTime;
+  }
+
+  public Optional<Program> getProgramByName(String name) {
+    return getPrograms().stream()
+        .filter(p -> p.getProgramDefinition().adminName().equals(name))
+        .findAny();
```

Comment:
Question (blocking): should this be `findOne`? What guarantees are there about the program this finds? Please document this and `getQuestionByName`

Reply:
Streams don't have findOne() - this is equivalent to that.

Reply:
ah bummer